### PR TITLE
Generate structs for event data.

### PR DESCRIPTION
**Pull request by nlordell**
This PR generates struct for event data instead of using tuples.

The generated code around anonymous events and events with anonymous fields is not well tested. I created issue #245 to add an example and test for this situation.

### Test Plan

Example uses generated type.

### CODE REVIEW

File: generate/src/contract/events.rs [hunk 1]
```diff
@@ -1,21 +1,149 @@
...
 
+fn expand_structs_mod(cx: &Context) -> Result<TokenStream> {
+    let structs = cx
+        .artifact
+        .abi
+        .events()
+        .map(|event| expand_struct(event))
+        .collect::<Result<Vec<_>>>()?;
+    if structs.is_empty() {
+        return Ok(quote! {});
```
Comment (e00E) [inline generate/src/contract/events.rs:31]: If I understand this correctly we only create the `pub mod events` if it is not empty. Why is that better than having the empty module? Not saying it is wrong, just curious.

Comment (nlordell) [reply generate/src/contract/events.rs:31]: Um, I wouldn't say its better, just an arbitrary decision I made. Let me know if you guys would prefer an empty `events` module.

File: generate/src/contract/events.rs [hunk 2]
```diff
@@ -1,21 +1,149 @@
...
+
+    let param_tokens = params
+        .iter()
+        .map(|(name, ty)| {
+            quote! {
+                let #name = #ty::from_token(tokens.next().unwrap())?;
+            }
+        })
+        .collect::<Vec<_>>();
+    let param_len = Literal::usize_unsuffixed(params.len());
```
Comment (e00E) [inline generate/src/contract/events.rs]: This part is hard to follow for me. Could we split it up into smaller functions and/or add tests? Don't do it if you think it's fine.

**Review by e00E — APPROVED** [review]

File: generate/src/contract/events.rs (hunk 1 above)
Comment (e00E) [reply generate/src/contract/events.rs:31]: The only reason I can think of is that someone might always do something like `use contract::events::*` (maybe this also happens in a macro). Then this would fail if the module didn't exist but it would be fine if it was empty.
On the other hand it could be confusing that there is an empty module, too.
I am fine with either.

File: generate/src/contract/events.rs (hunk 2 above)
Comment (nlordell) [reply generate/src/contract/events.rs]: @e00E let me know if its better now.

File: generate/src/contract/events.rs (hunk 1 above)
Comment (nlordell) [reply generate/src/contract/events.rs:31]: I'm not sure either. Maybe @fleupold has an opinion. One advantage of not generating `events` module is that if something assumes its there but it isn't then you get a compiler error. I usually prefer to err on the side of caution.

File: generate/src/contract/events.rs (hunk 2 above)
Comment (e00E) [reply generate/src/contract/events.rs]: Yes, thanks.

File: generate/src/contract/events.rs (hunk 1 above)
Comment (fleupold) [reply generate/src/contract/events.rs:31]: No strong feelings either way (let's be consistent how we handle this for other things such as methods). 
I'd argue you would get a compile error anyways when trying to access anything inside the empty module and just using a module without accessing it is not really useful (and will probably be a clippy warning to start with)

File: generate/src/contract/events.rs [hunk 3]
```diff
@@ -1,21 +1,189 @@
...
+        impl #event_name {
+            /// Retrieves the signature for the event this data corresponds to.
+            /// This signature is the Keccak-256 hash of the ABI signature of
+            /// this event.
+            pub fn signature() -> self::ethcontract::H256 {
+                #signature
+            }
+
+            /// Retrieves the ABI signature for the event this data corresponds
+            /// to.
```
Comment (fleupold) [inline generate/src/contract/events.rs]: bit: maybe include example for Transfer?

File: generate/src/contract/events.rs [hunk 4]
```diff
@@ -70,6 +235,8 @@ fn expand_filter(event: &Event) -> Result<TokenStream> {
     })
 }
 
+/// Expands into the `all_events` method on the root contract type if it
+/// contains events. Expands to nothing otherwise.
```
Comment (fleupold) [inline generate/src/contract/events.rs:244]: So all_events will continue to return events untyped? Not int his PR, but does it make sense to autogenerate an overarching all events enum that can have each event as a variant?

File: generate/src/contract/events.rs [hunk 5]
```diff
@@ -1,21 +1,189 @@
...
+fn expand_data_type(event: &Event) -> Result<TokenStream> {
+    let event_name = expand_struct_name(event);
+
+    let signature = expand_hash(event.signature());
+    let abi_signature = Literal::string(&event.abi_signature());
+
+    let params = expand_params(event)?;
+
+    let all_anonymous_fields = event.inputs.iter().all(|input| input.name.is_empty());
+    let (data_type_def, data_type_cstr) = if all_anonymous_fields {
```
Comment (fleupold) [inline generate/src/contract/events.rs]: ubernit: maybe data_type_definition/construction instead of abbreviation (cstr reminds me of C++)

File: generate/src/contract/events.rs [hunk 6]
```diff
@@ -1,21 +1,189 @@
...
+
+/// Expands an ABI event into name-type pairs for each of its parameters.
+fn expand_params(event: &Event) -> Result<Vec<(TokenStream, TokenStream)>> {
+    event
+        .inputs
+        .iter()
+        .enumerate()
+        .map(|(i, input)| {
+            // NOTE: Events can contain nameless values.
+            let name = methods::expand_input_name(i, &input.name);
```
Comment (fleupold) [inline generate/src/contract/events.rs]: nit: maybe lift from methods to util?

**Review by fleupold — APPROVED** [review]
You make it all seem so easy! 🍦

File: generate/src/contract/events.rs (hunk 4 above)
Comment (nlordell) [reply generate/src/contract/events.rs:244]: Yes, we will generate an enum with all even variants (including anonymous events) do it that way. Currently the runtime (the `ethcontract::contract::*` stuff) doesn't quite support it. Its definitely coming next.

File: generate/src/contract/events.rs (hunk 5 above)
Comment (nlordell) [reply generate/src/contract/events.rs]: Hehe

File: generate/src/contract/events.rs (hunk 1 above)
Comment (nlordell) [reply generate/src/contract/events.rs:31]: RIght now we omit things that would be empty, so I will keep as is then.

File: generate/src/contract/events.rs (hunk 3 above)
Comment (nlordell) [reply generate/src/contract/events.rs]: Even better, I generate the signature and inline it into the documentation :birthday:

File: generate/src/contract/events.rs [hunk 7]
```diff
@@ -132,19 +304,76 @@ mod tests {
...
+                    kind: ParamType::Bool,
+                    indexed: false,
+                },
+                EventParam {
+                    name: String::new(),
+                    kind: ParamType::Address,
+                    indexed: false,
+                },
+            ],
+            anonymous: false,
```
Comment (bh2smith) [inline generate/src/contract/events.rs:361]: What is this "anonymous" event component I have never heard of until now?

File: generate/src/contract/events.rs [hunk 8]
```diff
@@ -132,19 +304,76 @@ mod tests {
...
+        let (definition, construction) = expand_data_tuple(&name, &params);
+
+        assert_quote!(definition, {
+            struct Foo(pub bool, pub self::ethcontract::Address);
+        });
+        assert_quote!(construction, { Foo(p0, p1) });
+    }
+
+    #[test]
+    #[rustfmt::skip]
```
Comment (bh2smith) [inline generate/src/contract/events.rs]: Is it the extremely long string here that is being skipped?

File: generate/src/util.rs [hunk 9]
```diff
@@ -18,6 +19,17 @@ pub fn safe_ident(name: &str) -> Ident {
     syn::parse_str::<SynIdent>(name).unwrap_or_else(|_| ident(&format!("{}_", name)))
 }
 
+/// Expands a positional identifier string that may b
```
Comment (bh2smith) [inline generate/src/util.rs:22]: may b what?

File: generate/src/util.rs [hunk 10]
```diff
@@ -60,6 +72,21 @@ pub fn http_get(url: &str) -> Result<String> {
...
+
+    #[test]
+    fn input_name_to_ident_keyword() {
+        assert_quote!(expand_input_name(0, "self"), { self_ });
+    }
+
+    #[test]
+    fn input_name_to_ident_snake_case() {
+        assert_quote!(expand_input_name(0, "CamelCase1"), { camel_case_1 });
+    }
```
Comment (bh2smith) [inline generate/src/util.rs:88]: Why does self wind up with an underscore and not CamelCase1?

File: generate/src/contract/events.rs (hunk 7 above)
Comment (nlordell) [reply generate/src/contract/events.rs:361]: Solidity supports anonymous events. I also had no idea until I started doing some research for this feature.

Comment (nlordell) [reply generate/src/contract/events.rs:361]: Anonymous events don't include the event signature in topic 0, they basically just put the first indexed field into topic 0, so you lose the ability to filter on specific kinds of events from an event signature. The advantage is you get one extra topic to filter on.

I imagine this can be useful for contracts that only have a single event.

File: generate/src/util.rs (hunk 10 above)
Comment (nlordell) [reply generate/src/util.rs:88]: `self` is a keyword. As a workaround, we append `_` to it so that there aren't compiler errors. From the doc comments:
```
/// Expands an identifier string into a token and appending `_` if the
/// identifier is for a reserved keyword.
///
/// Parsing keywords like `self` can fail, in this case we add an underscore.
```
